tls/s2n_server_new_session_ticket.c

```diff
@@ -74,3 +75,75 @@ int s2n_server_nst_send(struct s2n_connection *conn)
 
     return 0;
 }
+
+int s2n_tls13_server_nst_send(struct s2n_connection *conn)
+{
+    notnull_check(conn);
+
+    /* Write message type because session resumption in TLS13 is a post-handshake message */
+    GUARD(s2n_stuffer_write_uint8(&conn->handshake.io, TLS_SERVER_NEW_SESSION_TICKET));
+
+    struct s2n_stuffer_reservation message_size = { 0 };
+    GUARD(s2n_stuffer_reserve_uint24(&conn->handshake.io, &message_size));
+
+    /* Ticket lifetime is the minimum between key lifetime, session state lifetime, and one week */
+    uint32_t key_lifetime_in_secs = (conn->config->encrypt_decrypt_key_lifetime_in_nanos +
+            conn->config->decrypt_key_lifetime_in_nanos) / ONE_SEC_IN_NANOS;
+    uint32_t session_state_lifetime_in_secs = conn->config->session_state_lifetime_in_nanos / ONE_SEC_IN_NANOS;
+
+    uint32_t minimum_key_and_session_lifetime = MIN(key_lifetime_in_secs, session_state_lifetime_in_secs);
+    uint32_t ticket_lifetime = MIN(minimum_key_and_session_lifetime, ONE_WEEK_IN_SECS);
+
+    GUARD(s2n_stuffer_write_uint32(&conn->handshake.io, ticket_lifetime));
+
+    /* Get random data to use as ticket_age_add value */
+    struct s2n_ticket_fields ticket_fields = { 0 };
+    struct s2n_blob random_data = { 0 };
+    GUARD(s2n_blob_init(&random_data, ticket_fields.ticket_age_add, sizeof(ticket_fields.ticket_age_add)));
+    GUARD_AS_POSIX(s2n_get_private_random_data(&random_data));
+    GUARD(s2n_stuffer_write_bytes(&conn->handshake.io, random_data.data, sizeof(ticket_fields.ticket_age_add)));
+
+    /* Write ticket nonce */
+    GUARD(s2n_stuffer_write_uint8(&conn->handshake.io, sizeof(conn->tickets_sent)));
+    GUARD(s2n_stuffer_write_uint16(&conn->handshake.io, conn->tickets_sent));
+
+    /* Derive individual session ticket secret */
+    s2n_tls13_connection_keys(secrets, conn);
+    s2n_stack_blob(ticket_secret, secrets.size, S2N_TLS13_SECRET_MAX_LEN);
+
+    uint8_t blob_data[sizeof(uint16_t)] = { 0 };
+    struct s2n_blob nonce_blob = { 0 };
+    struct s2n_stuffer ticket_nonce = { 0 };
+    GUARD(s2n_blob_init(&nonce_blob, blob_data, sizeof(conn->tickets_sent)));
+    GUARD(s2n_stuffer_init(&ticket_nonce, &nonce_blob));
+    GUARD(s2n_stuffer_write_uint16(&ticket_nonce, conn->tickets_sent));
```

Comment:
Is this necessary? If you don't use the stuffer "network order handling", do you end up with the bytes in the wrong order?

Reply:
When we're reading the ticket, won't we read back a byte array instead of a uint16? So it shouldn't actually matter what order the nonce bytes are in, they just need to be unique.